### PR TITLE
Newlines v2

### PR DESCRIPTION
**Pull request by mmmries**
A fresh attempt to support newlines. This pull request will hopefully fix #3 without the issues that the previous pull request had.

### CODE REVIEW

**Comment by mmmries** [comment]
@paulcsmith this is a bit of a rough draft, but it is passing the basic specs and I like it a whole lot better than my first pass. Any comments from you?

**Comment by catalystmediastudios** [comment]
I'll take a look as soon as I get home! I glanced over it and like it! I just have a couple comments.

File: lib/docx/newline_replacer.rb [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    def replace_text_node(parent, node)
+      text = node.to_s
+      list = text.split("\n")
+        .map{|str| str_to_text_node(str)}
+        .flat_map{ |txt| [txt,br] }
+      list.pop #remove trailing br
+      list.reverse.each do |new_node|
+        parent.insert_after(node, new_node)
+      end
+      node.remove
```
Comment (paulcsmith) [inline lib/docx/newline_replacer.rb:32]: Could we break this into smaller private methods that clarify what each line does?

File: lib/docx/newline_replacer.rb [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    end
+
+    def br
```
Comment (paulcsmith) [inline lib/docx/newline_replacer.rb]: What do you think about naming this something like `line_break_node` or `newline_node`? I think that might make it's use more clear and make it easier to understand in when used in `replace_text_node`.

**Comment by paulcsmith** [comment]
I left a couple comments. I mostly think some of the methods could be split apart to make it a bit easier to understand. Other than that, I like how it looks!

File: lib/docx/newline_replacer.rb [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+
+    def walk(node)
+      node.children.each do |child|
+        if child.node_type == :text
+          replace_text_node(node,child)
+        else
+          walk(child)
+        end
+      end
+    end
```
Comment (mmmries) [inline lib/docx/newline_replacer.rb:26]: I don't like this method very much. The if-else combined with an iteration seems a bit wonky for some reason, but I don't have a better idea off hand.

File: lib/docx/newline_replacer.rb [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+    def line_break_node
+      br = REXML::Element.new('w:br')
+      br.add_attribute('w:type', 'text-wrapping')
+      br
+    end
+
+    def list_of_new_nodes(node)
+       node.to_s.split("\n")
+        .map{|str| str_to_text_node(str)}
```
Comment (mmmries) [inline lib/docx/newline_replacer.rb:41]: Wondering if I should inline the str_to_text_node(str) call, now that this is a separate method just for generating a list of nodes it seems a little redundant to be calling a method which just builds a node

**Comment by mmmries** [comment]
I did a few quick refactorings, but left some additional notes of parts of the code I'm not :smiley_cat: about.

It might just be good enough for now :clock10: but if you have any other refactorings in mind feel free to take a stab, otherwise I'll probably bump the gem version and release it on Monday

File: lib/docx/newline_replacer.rb (hunk 4 above)
Comment (paulcsmith) [inline lib/docx/newline_replacer.rb:41]: I kind of like it actually, it makes it a bit more clear what's going on without muddying it up with how it's doing it

Comment (paulcsmith) [inline lib/docx/newline_replacer.rb:41]: Though if you did decide to inline it, I think that would be just fine as well

File: lib/docx/newline_replacer.rb [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    def replace
+      walk(doc.root)
+    end
+
+    private
+    attr_reader :doc
+
+    def walk(node)
+      node.children.each do |child|
+        replace_text_node_or_continue_walking(node, child)
```
Comment (paulcsmith) [inline lib/docx/newline_replacer.rb:16]: What do you think of this?

File: lib/docx/newline_replacer.rb [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+      if child.node_type == :text
+        replace_text_node(node, child)
+      else
+        walk(child)
+      end
+    end
+
+    def replace_text_node(parent, node)
+      list_of_new_nodes(node).reverse.each do |new_node|
+        parent.insert_after(node, new_node)
```
Comment (paulcsmith) [inline lib/docx/newline_replacer.rb:30]: As I was going through the code some more, I realized I'm not exactly sure what this is supposed to do? What is the "new_node" in this case. Is it a node with a line break? I'm thinking the names could be clarified. What do you think?

**Comment by paulcsmith** [comment]
I made another comment and a small refactor. I think it (clarification: the whole PR) looks really good, though I wouldn't mind spending some more time now and cleaning things up a bit further. Totally up to you :)

File: lib/docx/newline_replacer.rb (hunk 6 above)
Comment (mmmries) [inline lib/docx/newline_replacer.rb:30]: So this method basically splits up a text node into a list of new nodes and then inserts them into the parent node and removes the original text node.

So if the original XML looks like this:

``` xml
<node>Some text with\nnewlines\naround</node>
```

We would take that single text node and turn it into three text nodes with two break nodes between them:

``` xml
<node>Some text with<w:br/>newlines<w:br/>around</node>
```

**Comment by mmmries** [comment]
I'm find to take some time to do cleanup. I just don't want to forget to release it. What do you think of doing a 0.1.3 release from this branch and then opening a new PR with some refactorings?

IMO I think refactoring other parts of the codebase would be more beneficial than refactoring this section of the code (famous last words).

**Comment by leducmarc** [comment]
Was going to test this out but after switching to :branch => 'newlines_v2' now the \n's aren't creating new lines anymore. Is it a different markup for them now in our strings? Am I missing something obvious? xD

**Comment by mmmries** [comment]
@giantchiprel just make sure that when you create your DocxTemplater object you pass in the convert_newlines option.  ie:

``` ruby
DocxTemplater.new(convert_newlines: true).replace_file_with_content('path/to/file.docx')
```

**Comment by paulcsmith** [comment]
@hqmq Yeah it's probably good to release and we can refactor other parts of the code later :) We may want to update the README as well though

**Comment by leducmarc** [comment]
Cool, so it was that parameter I was missing and it's working now!

Seem to get a new glitch that wasn't happening before though, it doesn't like my apostrophes =P

i.e. 
:cake => "Cake can't eat you"
outputs "Cake can&apos;t eat you"

I can get around it using the other version, but probably not ideal
example instead of can't I put can’t (The apostrophe is different, english vs french versions or something?)

**Comment by mmmries** [comment]
@paulcsmith good call! We definitely should mention it in the README befor I merge and release

@giantchiprel I'll do some manual testing of the apostrophe issue tonight and see if I can reproduce. If I can reproduce it I'll throw in a test to see if we can get it fixed as part of the same release.

**Comment by leducmarc** [comment]
Also, it looks like the way you're doing it you aren't far from allowing real carriage returns (new paragraphs) on top of just line breaks. Is that feasible for the future? It ends up mattering for formatting, example being 'justify text' where it stretches the spacing for the phrase to fill the full line EXCEPT the last line of a paragraph, so \n stretches the last line to fit the whole thing regardless of length since a line break isn't the end of a paragraph.

From what it looks like, you'd have to find the BR and insert a close-line/close-paragraph/open-line/open-paragraph to accomplish a true paragraph; Not sure if it's feasible but MIGHT be!

**Comment by leducmarc** [comment]
@hqmq Great, thanks!

Oh, looks like the forum turned the output into proper formatting xD
it outputs: "Cake can&*apos;t eat you" (Without the *)

**Comment by leducmarc** [comment]
So in the raw word-processingML if I insert

```
</w:r></w:p><w:p><w:r><w:t>
```

it will create a line break and start a new paragraph. However the formatting will be lost as I didn't include a <w:rPr> block with formatting between the p and r. Technically it should be the same formatting as it previously was, so ideally it would snatch the last declared formatting block and repeat it, but that's obviously not a priority.

Would something like this be feasible to implement using the code already there for \n? Like a \p command that starts a new paragraph instead of just a line break.

Cheers!
Marc

**Comment by mmmries** [comment]
@giantchiprel I found and fixed the issue with apostrophes (thanks for finding it :smiley_cat:)

I think trying to do full paragraph breaking is a bit outside the scope of this pull request. The problem that I see is the context. We might encounter a newline in the middle of a replacement that is meant to go into a table or a paragraph with lots of styling and if we just break the xml in the middle it could break a lot of structure in the document.

I think that if you want to affect the structure of a document you should probably look for a library that builds up the structure as it goes, rather than attempting to replace structure in an existing document. Perhaps something like https://github.com/docxtor/docxtor might fit that need better.

**Comment by mmmries** [comment]
@paulcsmith does the README look alright? Are you okay with pushing the new version?

File: README.md [hunk 7]
```diff
@@ -22,6 +22,18 @@ Example usage:
      File.open("path/to/mydocument.docx", "wb") {|f| f.write(buffer.string) }
 ```
 
+Newline Conversion
+==================
+
+Sometimes of one of your replacement values is meant to span multiple lines it can be useful to convert newlines into line breaks in the docx file.
```
Comment (paulcsmith) [inline README.md]: There is an extra "of" near the beginning

File: README.md [hunk 8]
```diff
@@ -22,6 +22,18 @@ Example usage:
...
+
+This can be accomplishes with the :convert_newlines option like this:
```
Comment (paulcsmith) [inline README.md]: Accomplishes -> accomplished

**Comment by paulcsmith** [comment]
Yep I'm ok with pushing it and the read me looks great. Just two small comments. 

I think it's good to merge though I have one thought. Should the default be to convert newlines? I'm beginning to think that would be the desired behavior most of the time right? What do you think?

**Comment by mmmries** [comment]
The only reason I defaulted this behavior to off was for backwards compatability (ie in case someone was already using the gem and their code was happily ignoring newlines until they update). But that seems pretty unlikely. There are only a bit over 1,000 downloads of the gem so maybe we should go ahead and make this a default behavior that can be turned off.

I can switch around the README and the tests so that it defaults on.

**Comment by mmmries** [comment]
@paulcsmith I've switched up the default and fixed the README.

I also found one bug relating to preserving whitespace during newline replacement.

Any final veto?

**Comment by paulcsmith** [comment]
Looks awesome @hqmq! Looks good to merge

**Comment by leducmarc** [comment]
So I'm still new to all this github'ing, but I forked the project and added support for new paragraphs, which I may extend later for extra customizability (i.e. add elements to *Pr elements in p and/or r for formatting, can be put in as parametrized text in tags) but for now it serves my purpose. Figured I'd mention it here incase you guys wanted to have a look or take parts or however this github works! xD

For the record, I addressed the issues pointed out above by Xpath searching for paragraphs regardless of context and manipulated the structure from there with a recursive loop which repeats previous formatting where needed. I don't feel the need for performance atm so I just patched it on post-replacesment but pre-new lines, brute force replicating everything with text instead of actually using a form of smart/selective search. End result, slightly slower, but seems to work as intended.

Any comments are more than welcome and appreciated; Thanks again for holding my hand a little on my first séjour into the github ;p

**Pull request merged**